File: test/benchmark/ruby/binary_trees.rb

class Tree
    attr_accessor :depth, :item, :left, :right
    def initialize(item, depth)
        @item = item
        @depth = depth
        if depth > 0
            item2 = item + item
            depth = depth - 1
            @left = Tree.new(item2 - 1, depth)
            @right = Tree.new(item2, depth)
        else
            @left = nil
            @right = nil
        end
    end

    def check()
        if @left == nil
            return @item
        end

        return @item + @left.check() - @right.check()
    end
end


minDepth = 4
maxDepth = 14
stretchDepth = maxDepth + 1

start = Time.now

puts("stretch tree of depth:")
p(stretchDepth)
puts("check:")
p(Tree.new(0, stretchDepth).check())

longLivedTree = Tree.new(0, maxDepth)

# iterations = 2 ** maxDepth
iterations = 1
d = 0
while d < maxDepth
    iterations = iterations * 2
    d = d + 1
end

depth = minDepth
while depth < stretchDepth
    check = 0
    i = 1
    while i <= iterations
        check = check + Tree.new(i, depth).check() + Tree.new(-i, depth).check()
        i = i + 1
    end

    puts("num trees:")
    p(iterations * 2)
    puts("depth:")
    p(depth)
    puts("check:")
    p(check)

    iterations = iterations / 4
    depth = depth + 2
end

puts("long lived tree of depth:")
p(maxDepth)
puts("check:")
p(longLivedTree.check())
puts("elapsed:")
p(Time.now - start)

File: test/benchmark/python/binary_trees.py
import time


class Tree:
    def __init__(self, item, depth):
        self.item = item
        self.depth = depth
        if depth > 0:
            item2 = item + item
            depth = depth - 1
            self.left = Tree(item2 - 1, depth)
            self.right = Tree(item2, depth)
        else:
            self.left = None
            self.right = None

    def check(self):
        if self.left == None:
            return self.item

        return self.item + self.left.check() - self.right.check()


minDepth = 4
maxDepth = 14
stretchDepth = maxDepth + 1

start = time.process_time()

print("stretch tree of depth:")
print(stretchDepth)
print("check:")
print(Tree(0, stretchDepth).check())

longLivedTree = Tree(0, maxDepth)

# iterations = 2 ** maxDepth
iterations = 1
d = 0
while d < maxDepth:
    iterations = iterations * 2
    d = d + 1

depth = minDepth
while depth < stretchDepth:
    check = 0
    i = 1
    while i <= iterations:
        check = check + Tree(i, depth).check() + Tree(-i, depth).check()
        i = i + 1

    print("num trees:")
    print(iterations * 2)
    print("depth:")
    print(depth)
    print("check:")
    print(check)

    iterations = iterations / 4
    depth = depth + 2

print("long lived tree of depth:")
print(maxDepth)
print("check:")
print(longLivedTree.check())
print("elapsed:")
print(time.process_time() - start)
